hogben/tests/test_optimise.py

```diff
@@ -0,0 +1,134 @@
+    num_contrasts = 3
+    angle_times = [(0.7, 100, 10), (2.3, 100, 40)]
+
+    # Mock values retreived from previous run
+    mock_optimise.return_value = (
+        np.array([-0.56, 2.15, 6.36, 0.17, 0.28, 0.56]), -0.18
+    )
+    contrasts, splits, _ = optimiser.optimise_contrasts(num_contrasts,
+                                                        angle_times,
+                                                        workers=-1,
+                                                        verbose=False)
+    assert len(contrasts) == num_contrasts and len(splits) == num_contrasts
+
+
+@patch('hogben.optimise.Optimiser._Optimiser__optimise')
+def test_optimise_underlayers(mock_optimise):
+    """
+    Tests that the optimise_contrasts method outputs the correct amount of
+    contrasts and counting times.
+    """
+    optimiser = Optimiser(BilayerDMPC())
+
+    num_underlayers = 3
+    angle_times = [(0.7, 100, 10), (2.3, 100, 40)]
+    contrasts = [-0.56, 6.36]
+    thick_bounds = (0, 500)
+    sld_bounds = (1, 9)
+
+    # Mock values retreived from previous run
+    mock_optimise.return_value = (
+        np.array([-0.56, 2.15, 6.36, 0.17, 0.28, 0.56]), -0.18
+    )
+
+    contrasts, splits, _ = optimiser.optimise_underlayers(num_underlayers,
+                                                          angle_times,
+                                                          contrasts,
+                                                          thick_bounds,
+                                                          sld_bounds,
+                                                          verbose=False)
+    assert len(contrasts) == num_underlayers and len(splits) == num_underlayers
+
+
+def test_angle_times_func_result(refnx_sample):
+    """Checks that the angle_times_func method gives the correct result"""
+    x = [0.3, 1.3, 0.8, 0.2]  # [angle, angle, time, time]
+    num_angles = 2
+    contrasts = [3, 14, -2]
+    points = 100
+    total_time = 10000
+
+    optimiser = Optimiser(refnx_sample)
+    result = optimiser._angle_times_func(x, num_angles, contrasts, points,
+                                         total_time)
+    expected_result = -1.7721879778537162
+    np.testing.assert_allclose(result, expected_result, rtol=1e-06)
+
+
+def test_contrasts_func_result():
+    """Checks that the _contrasts_func method gives the correct result"""
+    x = [0.3, 9.3, 0.8, 8.2]  # [SLD, SLD, time, time]
+    num_contrasts = 2
+    angle_splits = [(0.7, 100, 0.6), (2.3, 100, 0.4)]
+    total_time = 100000
+
+    optimiser = Optimiser(BilayerDMPC())
+    result = optimiser._contrasts_func(x, num_contrasts, angle_splits,
+                                       total_time)
+    expected_result = -2.557881096447291
+    np.testing.assert_allclose(result, expected_result, rtol=1e-06)
+
+
+def test_underlayers_func():
+    """Checks that the _underlayers_func method gives the correct result"""
+    x = [50, 20, -5, 10]  # [thickness, thickness, SLD, SLD]
```

Comment:
even in tests I'm not a fan of using `x` as a variable name

Reply:
The `x` as variable name is simply what the tested function uses as well. But I agree that it's a poor choice of naming, which maybe is even apparant from the fact that I needed to add a clarifying comment after the assignments. 

Will ammend these names.